_LowLevelCode/cpp/cpp_communicator/MPI_wrapper.cpp

```diff
@@ -573,6 +632,53 @@ void MPI_wrapper::clearAll() {
     }
 }
 
+/** pack node_names variable into linear buffer to be able to send it over MPI with one message
+ and restore it later (kind of primitive serialization)
+ Input:
+     takes the structure of the class variable node_names
+ Output:
+    fills in the buffer vector with the restorable structure values
+*/
+void MPI_wrapper::pack_node_names_list(std::vector<char>& buf)const {
+
+    size_t i;
+    size_t BlockSize = sizeof(size_t);
+    // estimate the size of the nodenames data
+    size_t buf_size = BlockSize;
+    for (i = 0; i < this->node_names.size(); i++) {
+        buf_size += (this->node_names[i].size() + 1);
+    }
+    // fill-in the buffer
+    buf.reserve(buf_size);
+    size_t sizebuf = this->node_names.size();
+    auto pBuf = reinterpret_cast<char*>(&sizebuf);
+    for (i = 0; i < BlockSize; i++) buf.push_back(pBuf[i]);
+
+    // copy all node names
+    for (i = 0; i < this->node_names.size(); i++) {
+        for (size_t j = 0; j <= this->node_names[i].size(); j++) {
+            auto pString_start = this->node_names[i].c_str();
+            buf.push_back(pString_start[j]);
+        }
```

Comment:
Could you not just `append` the node_names, rather than `push_back` char by char?

Reply:
I am filing vector with null-terminated c_str(). Vector does not have `append` operation. Its double loop so `i` is upper and `j` is lower index. Filling string instead of vector with contents is possible but I base deserialization on null-terminated strings so gaining in simplicity here I will loose there.

Reply:
Chris' comment was to change it to:
```c
    for (i = 0; i < this->node_names.size(); i++) {
        auto pString_start = this->node_names[i].c_str(); // Does not depend on j
        for (size_t j = 0; j <= this->node_names[i].size(); j++) {
            buf.push_back(pString_start[j]);
        }
    }
```

Reply:
its possible though the modern compilers to this very well themselves.